syntax = "proto3";

package chat;

service ChatService {
  rpc CreateAccount(CreateAccountRequest) returns (Response) {}
  rpc Login(LoginRequest) returns (LoginResponse) {}
  rpc ListAccounts(ListAccountsRequest) returns (ListAccountsResponse) {}
  rpc SendMessage(SendMessageRequest) returns (Response) {}
  rpc ReadMessages(ReadMessagesRequest) returns (ReadMessagesResponse) {}
  rpc DeleteMessages(DeleteMessagesRequest) returns (Response) {}
  rpc DeleteAccount(DeleteAccountRequest) returns (Response) {}

  rpc MessageStream(MessageStreamRequest) returns (stream NewMessage) {}
}

message CreateAccountRequest {
  string username = 1;
  string password = 2;
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message ListAccountsRequest {
  string pattern = 1;
}

message SendMessageRequest {
  string sender = 1;
  string recipient = 2;
  string message_text = 3;
}

message ReadMessagesRequest {
  string username = 1;
  int32 num = 2;
}

message DeleteMessagesRequest {
  string username = 1;
  repeated string message_ids = 2;
}

message DeleteAccountRequest {
  string username = 1;
  string password = 2;
}

message MessageStreamRequest {
  string username = 1;
}

message Response {
  string status = 1;
  string message = 2;
}

message LoginResponse {
  string status = 1;
  string message = 2;
  int32 unread_count = 3;
}

message ListAccountsResponse {
  string status = 1;
  string message = 2;
  repeated string accounts = 3;
}

message ReadMessagesResponse {
  string status = 1;
  string message = 2;
  repeated string messages = 3;
}

message NewMessage {
  string sender = 1;
  string message_text = 2;
  string msg_id = 3;
}
